testsuite/MDAnalysisTests/coordinates/test_h5md.py

```diff
@@ -0,0 +1,470 @@
+                f.copy(source='h5md', dest=g)
+                g['particles/trajectory/velocity/value'].resize((3, 10000, 3))
+        with pytest.raises(ValueError):
+            u = mda.Universe(TPR_xvf, outfile, format='H5MD')
+
+
+@pytest.mark.skipif(not HAS_H5PY, reason="h5py not installed")
+def test_changing_n_atoms3(h5md_file, ref, tmpdir):
+    outfile = 'test_changing_n_atoms3' + ref.ext
+    with tmpdir.as_cwd():
+        with h5md_file as f:
+            with h5py.File(outfile, 'w') as g:
+                f.copy(source='particles', dest=g)
+                f.copy(source='h5md', dest=g)
+                g['particles/trajectory/force/value'].resize((3, 10000, 3))
+        with pytest.raises(ValueError):
+            u = mda.Universe(TPR_xvf, outfile, format='H5MD')
+
+@pytest.mark.skipif(not HAS_H5PY, reason="h5py not installed")
+def test_2D_box(h5md_file, ref, tmpdir):
+    outfile = 'test_2D_box' + ref.ext
+    with tmpdir.as_cwd():
+        with h5md_file as f:
+            with h5py.File(outfile, 'w') as g:
+                f.copy(source='particles', dest=g)
+                f.copy(source='h5md', dest=g)
+                new_box = np.ones(shape=(3,2,2))
+                g['particles/trajectory/box'].attrs['dimension'] = 2
+                del g['particles/trajectory/box/edges/value']
+                g['particles/trajectory'
+                  '/box/edges'].create_dataset('value', data=new_box)
+        with pytest.raises(ValueError):
+            u = mda.Universe(TPR_xvf, outfile, format='H5MD')
+
+@pytest.mark.skipif(not HAS_H5PY, reason="h5py not installed")
+def test_no_box(h5md_file, ref, tmpdir):
+    outfile = 'test_no_box' + ref.ext
+    with tmpdir.as_cwd():
+        with h5md_file as f:
+            with h5py.File(outfile, 'w') as g:
+                f.copy(source='particles', dest=g)
+                f.copy(source='h5md', dest=g)
+                del g['particles/trajectory/box/edges']
+        u = mda.Universe(TPR_xvf, outfile, format='H5MD')
+        assert_array_equal(u.trajectory.ts.dimensions, np.zeros((6,)))
+
+@pytest.mark.skipif(not HAS_H5PY, reason="h5py not installed")
+def test_no_groups(h5md_file, ref, tmpdir):
+    outfile = 'test_no_groups' + ref.ext
+    from MDAnalysis.exceptions import NoDataError
+    with tmpdir.as_cwd():
+        with h5md_file as f:
+            with h5py.File(outfile, 'w') as g:
+                f.copy(source='particles', dest=g)
+                f.copy(source='h5md', dest=g)
+                del g['particles/trajectory/position']
+                del g['particles/trajectory/velocity']
+                del g['particles/trajectory/force']
+        with pytest.raises(NoDataError):
+            u = mda.Universe(TPR_xvf, outfile, format='H5MD')
+
+@pytest.mark.skipif(not HAS_H5PY, reason="h5py not installed")
+def test_open_filestream(h5md_file):
+    with h5md_file as f:
+        from MDAnalysis.lib.util import NamedStream
+        u = mda.Universe(TPR_xvf, NamedStream(f, f.filename))
```

Comment:
This should not need `NamedStream`. There's something wrong with how `_format_hint` is used (or not used), I think. At least my impression was,
```
u = mda.Universe(TPR_xvf, f)
```
should work.

Reply:
Thank you for adding a test!!

Reply:
Looking through now to see what `_format_hint` is doing

Reply:
I've found why it's not auto detecting the file format - 

in `_get_readers.py`:
https://github.com/MDAnalysis/mdanalysis/blob/2f16a16aa3b9484ba37c9281e32bc80a18e5878e/package/MDAnalysis/core/_get_readers.py#L91-L93

which gets passed to `guess_format()` in MDAnalysis.lib:
https://github.com/MDAnalysis/mdanalysis/blob/2f16a16aa3b9484ba37c9281e32bc80a18e5878e/package/MDAnalysis/lib/util.py#L992-L1000

The problem is that an `h5py.File` uses the `.name` attribute as the _group_ name, not filename:
![image](https://user-images.githubusercontent.com/59893899/89251570-36f08b00-d5cc-11ea-8197-0bdb64cc75cf.png)
so I don't think it can guess the file format from the extension.

Reply:
In the `_get_reader_for` documentation, it mentions 
https://github.com/MDAnalysis/mdanalysis/blob/2f16a16aa3b9484ba37c9281e32bc80a18e5878e/package/MDAnalysis/core/_get_readers.py#L57-L68

but I can't find the code giving _how_ it handles these cases